File: frontend/src/api/songs.ts
import type { Song, SongDetail, ScrapeResponse } from "../types/song";

const BASE = import.meta.env.VITE_API_URL ?? "/_/backend";

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${BASE}${path}`, init);
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    // Preserve structured detail (e.g. {code, message}) as-is
    const detail = err.detail ?? "Erreur inconnue";
    const message = typeof detail === "object" ? detail.message : detail;
    const error = new Error(message) as Error & { code?: string };
    if (typeof detail === "object") error.code = detail.code;
    throw error;
  }
  if (res.status === 204) return undefined as T;
  return res.json();
}

export const getSongs = () => request<Song[]>("/songs");

export const getSong = (id: number) => request<SongDetail>(`/songs/${id}`);

export const scrapeSong = (url: string) =>
  request<ScrapeResponse>("/songs/scrape", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ url }),
  });

export const songFromText = (
  text: string,
  sourceUrl: string,
  title?: string,
  artist?: string,
) =>
  request<ScrapeResponse>("/songs/from-text", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ text, source_url: sourceUrl, title, artist }),
  });

export const updateSong = (id: number, title: string, artist: string) =>
  request<SongDetail>(`/songs/${id}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ title, artist }),
  });

export const deleteSong = (id: number) =>
  request<void>(`/songs/${id}`, { method: "DELETE" });

File: backend/app/api/songs.py
from urllib.parse import urlparse
from fastapi import APIRouter, HTTPException, status
from ..schemas import (
    ScrapeRequest, PasteRequest, ScrapeResponse,
    SetTagsRequest, SongDetail, SongOut,
)
from ..crud import (
    get_songs, get_song, get_song_by_url, create_song,
    delete_song, set_song_tags,
)
from ..scraper.fetcher import fetch_page, fetch_ug_page
from ..scraper.parser import parse_song
from ..scraper.ug_parser import parse_ug_song
from ..scraper.exceptions import ScrapeError, ParseError

router = APIRouter()

ALLOWED_DOMAINS = {"boiteachansons.net", "ultimate-guitar.com", "tabs.ultimate-guitar.com"}


def _get_host(url: str) -> str:
    return urlparse(url).netloc.replace("www.", "")


def _validate_url(url: str) -> str:
    host = _get_host(url)
    if not any(host == d or host.endswith("." + d) for d in ALLOWED_DOMAINS):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Domaine non supporté : {host}. Utilise boiteachansons.net ou ultimate-guitar.com",
        )
    return host


@router.post("/scrape", response_model=ScrapeResponse, status_code=status.HTTP_201_CREATED)
def scrape_song(body: ScrapeRequest):
    host = _validate_url(body.url)
    is_ug = "ultimate-guitar" in host

    existing = get_song_by_url(body.url)
    if existing:
        return ScrapeResponse(song=SongDetail.model_validate(existing), already_existed=True)

    try:
        if is_ug:
            html, canonical_url = fetch_ug_page(body.url)
        else:
            html, canonical_url = fetch_page(body.url)
    except ScrapeError as e:
        err_msg = str(e)
        if "cloudflare" in err_msg.lower():
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail={"code": "ug_blocked", "message": err_msg},
            )
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=err_msg)

    existing = get_song_by_url(canonical_url)
    if existing:
        return ScrapeResponse(song=SongDetail.model_validate(existing), already_existed=True)

    try:
        data = parse_ug_song(html, canonical_url) if is_ug else parse_song(html, canonical_url)
    except ParseError as e:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(e))

    song = create_song(data)
    return ScrapeResponse(song=SongDetail.model_validate(song))


@router.post("/from-text", response_model=ScrapeResponse, status_code=status.HTTP_201_CREATED)
def song_from_text(body: PasteRequest):
    """Fallback : importer un tab collé manuellement (format UG ou texte libre)."""
    from ..scraper.ug_parser import _ug_to_chordpro
    from ..scraper.parser import SongData

    text = body.text.strip()
    if not text:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Texte vide")

    # Detect UG format by presence of [ch] tags
    if "[ch]" in text:
        content = _ug_to_chordpro(text)
    else:
        content = text  # assume already ChordPro or plain text

    # Extract title/artist from first lines if they look like metadata
    lines = text.splitlines()
    title = "Titre inconnu"
    artist = "Artiste inconnu"
    for line in lines[:5]:
        if line.lower().startswith("title:"):
            title = line.split(":", 1)[1].strip().title()
        elif line.lower().startswith("artist:"):
            artist = line.split(":", 1)[1].strip().title()

    if body.source_url:
        existing = get_song_by_url(body.source_url)
        if existing:
            return ScrapeResponse(song=SongDetail.model_validate(existing), already_existed=True)

    data = SongData(
        title=title, artist=artist, key=None, capo=0,
        content=content, source_url=body.source_url or f"paste://{hash(text)}",
    )
    song = create_song(data)
    return ScrapeResponse(song=SongDetail.model_validate(song))


@router.get("", response_model=list[SongOut])
def list_songs():
    return [SongOut.model_validate(s) for s in get_songs()]


@router.get("/{song_id}", response_model=SongDetail)
def get_song_detail(song_id: int):
    song = get_song(song_id)
    if not song:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Chanson introuvable")
    return SongDetail.model_validate(song)


@router.put("/{song_id}/tags", response_model=SongDetail)
def update_song_tags(song_id: int, body: SetTagsRequest):
    song = set_song_tags(song_id, body.tag_ids)
    if not song:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Chanson introuvable")
    return SongDetail.model_validate(song)


@router.delete("/{song_id}", status_code=status.HTTP_204_NO_CONTENT)
def remove_song(song_id: int):
    if not delete_song(song_id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Chanson introuvable")
